fix remaining active socket in state SYN-SENT

ikoveshnikov opened the pull request:
fix #751

krizhanovsky commented on tempesta_fw/sock.c:
```diff
@@ -1212,13 +1212,26 @@ ss_tx_action(void)
-			if (!ss_sock_live(sk)) {
+			switch (sk->sk_state) {
+			case TCP_ESTABLISHED:
+				__sk_close_locked(sk); /* paired with bh_lock_sock() */
```
too long line

keshonok commented on tempesta_fw/sock_srv.c:
```diff
@@ -181,6 +181,8 @@ tfw_sock_srv_connect_try(TfwSrvConn *srv_conn)
 	}
 
+	/* Link Tempesta with the socket. */
+	tfw_connection_link_to_sk((TfwConn *)srv_conn, sk);
```
Also, it's important here that when Tempesta is stopped, `ss_close_sync(conn->sk, true);` always pushes the request for a close to the work queue, so the closing in never synchronous. That's because Tempesta is stopped while in user context, and the condition `!in_softirq()` makes `__ss_close()` do just that. That way the case of `TCP_SYN_SENT` needs to be handled only in `ss_tx_action()`.

Perhaps, a note in the description of `tfw_sock_srv_disconnect()` would help to understand the intricacies of the behaviour better.

krizhanovsky commented on tempesta_fw/sock_srv.c:
```diff
@@ -416,13 +417,13 @@ tfw_sock_srv_disconnect(TfwConn *conn)
-		ret = ss_close_sync(conn->sk, true);
+	if (sk) {
+		sock_hold(sk);
+		tfw_connection_unlink_to_sk(conn);
```
Again, I'm not sure about correctness of the change. We had many crashes when various contexts resolve already dead `conn->sk`, so now we call `tfw_connection_unlink_to_sk()` on connection destructor only, when `tfw_connection_put()` ensures that there are no more users of the connection. Just removing the link could lead to inconsistencies.

ikoveshnikov commented on tempesta_fw/t/functional/flacky_net/test_sockets.py:
```diff
@@ -0,0 +1,147 @@
+"""All sockets must be closed after Tempesta shutdown"""
```
Style guide in Python is not so flexible like in C. Python community forces every developer to use the  official code style described in PEP-8 document https://www.python.org/dev/peps/pep-0008 . Indentation in Python matters a lot since it is used instead old good operator brackets, so the community decided to have the only right variant: 4 spaces, tabs and mixing tabs with spaces are not allowed. Even more the wrong indentation is equal to syntax error in Python 3. `pylint`, the Python linter tool, threats any indentation different from 4 spaces as warnings.

I agree with you, that the whole project must use the same code style. But we can literately do nothing here.

krizhanovsky replied:
Well, I don't care about Guido's point of view on how a beauty code must looks like. I just verified that tabs work find for Python3:

```python
$ cat tail_recursion.py 
#!/usr/bin/python3

def tail_factorial(total, n):
	if not n:
		return total
	return tail_factorial(total * n, n - 1)

def tail_factorial2(total, n):
	if not n:
		return total
	return (lambda : tail_factorial2(total * n, n - 1))

def tail_rec(total, n):
	g = tail_factorial2(total, n)
	while hasattr(g, '__call__'):
		total = total * tail_factorial

print(tail_factorial(1, 10))
$ ./tail_recursion.py 
3628800
```

The code used tabs 8 characters in width. It seems Pylint allows to specify what should it use for identation https://stackoverflow.com/questions/25303726/how-to-fix-pylint-wrong-hanging-indentation-and-pep8-e121

ikoveshnikov commented:
Changes since last review:
* rework of server connection lifetime. All resources attached to server connections will be released once linked socket is closed
* add vim modeline to python files to set correct indentation rules.

keshonok commented on tempesta_fw/sock_srv.c:
```diff
@@ -174,19 +184,15 @@ tfw_sock_srv_connect_try(TfwSrvConn *srv_conn)
 				r);
 		ss_close_sync(sk, false);
+		/* Drop failed connection. */
+		srv_conn->proto.hooks->connection_drop(sk);
```
It appears that this is not very clean. For instance, where's a call to `tfw_connection_unlink_from_sk()`? Also, calling `connection_drop()` is not needed when a connection is not live -- see `tfw_sock_srv_connect_failover()`. Also, doing it this way unnecessarily increments the stats count of disconnects.

krizhanovsky commented on tempesta_fw/t/functional/flacky_net/test_sockets.py:
```diff
@@ -0,0 +1,147 @@
+
+    def test_unreachable(self):
+        """Server does not exist and cannot be reached."""
+        self.check_before_start()
```
The comment is still not addressed.
